### PR TITLE
[shard-map] handle scalar residuals in post_process partial eval

### PR DESCRIPTION
**Pull request by mattjj**
This is the third time fixing one bug! Well, it was three different instances of the same conceptual issue: `shard_map`ped functions can't have scalar outputs (unless they're the same value for every function instance) because out_specs only let us express concatenation not stacking. So when autodiff of the body produces scalar residuals, we need to add a new axis to those scalars to be able to concatenate them in the output of the forward pass, and then remove that extra axis on the backward pass.

We had fixed that bug before in shard_map's partial eval rule, and then in its custom-policy partial eval rule. Here finally we're fixing it in the third and last place: in the post-process partial eval rule, which gets called when we differentiate with respect to a variable closed over by the `shard_map`ped function and none of its arguments.

### CODE REVIEW

File: jax/experimental/shard_map.py [hunk 1]
```diff
@@ -985,19 +987,20 @@ def out_names_transform(out_names):
...
-                  for v in jaxpr.constvars]
-  out_consts_ = [jax.lax.broadcast(x, (1,)) if scalar else x
-                 for x, scalar in zip(out_consts, which_scalar)]
+  jaxpr, out_consts = _promote_scalar_residuals_jaxpr(jaxpr, out_consts)
+  yield jaxpr, (out_pvals, out_consts, env)
+
+def _promote_scalar_residuals_jaxpr(jaxpr, res):
+  which = [isinstance(v.aval, core.ShapedArray) and not v.aval.shape
+           for v in jaxpr.constvars]
+  res_ = [jax.lax.broadcast(x, (1,)) if s else x for x, s in zip(res, which)]
```
Comment (froystig) [inline jax/experimental/shard_map.py:996]: Howcome we don't `broadcast` like this in `_add_singleton` (and then use it here too, in turn)?

**Review by froystig — APPROVED** [review]

File: jax/experimental/shard_map.py (hunk 1 above)
Comment (mattjj) [reply jax/experimental/shard_map.py:996]: I think here `x` might be of Python builtin numeric type, so we can't count on it having a `reshape` method, whereas `_add_singleton` is currently called in places where `x` is guaranteed to have a `reshape` method. We could probably unify these but... probably not worth much!

Comment (froystig) [reply jax/experimental/shard_map.py:996]: Yep, that makes sense.
